## 1. Assignment Description
In this assignment, you will be required to explore, segment, and cluster the neighborhoods in the city of Toronto.

For the Toronto neighborhood data, a Wikipedia page exists that has all the information we need to explore and cluster the neighborhoods in Toronto. You will be required to scrape the Wikipedia page and wrangle the data, clean it, and then read it into a pandas dataframe so that it is in a structured format like the New York dataset.

Once the data is in a structured format, you can replicate the analysis that we did to the New York City dataset to explore and cluster the neighborhoods in the city of Toronto.

## 2. Scrap content from wiki page
Import necessary packages.

In [63]:
import pandas as pd
from bs4 import BeautifulSoup
import requests
import numpy as np
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
from pandas.io.json import json_normalize  # tranform JSON file into a pandas dataframe

import folium # map rendering library

# import k-means from clustering stage
from sklearn.cluster import KMeans

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

### web scraping

In [68]:
source = requests.get("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M").text
soup = BeautifulSoup(source, 'lxml')

table = soup.find("table")
table_rows = table.tbody.find_all("tr")

res = []
for tr in table_rows:
    td = tr.find_all("td")
    row = [tr.text for tr in td]
    
    # Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
    if row != [] and row[1] != "Not assigned":
        # If a cell has a borough but a "Not assigned" neighborhood, then the neighborhood will be the same as the borough.
        if "Not assigned" in row[2]: 
            row[2] = row[1]
        res.append(row)

# Dataframe with 3 columns
df = pd.DataFrame(res, columns = ["PostalCode", "Borough", "Neighborhood"])
df.head()

PostalCode             Borough                  Neighborhood
0      M1A\n      Not assigned\n                            \n
1      M2A\n      Not assigned\n                            \n
2      M3A\n        North York\n                   Parkwoods\n
3      M4A\n        North York\n            Victoria Village\n
4      M5A\n  Downtown Toronto\n  Regent Park / Harbourfront\n

Remove "\n" and replace with " "

In [69]:
df["PostalCode"] = df["PostalCode"].str.replace("\n","")
df["Borough"] = df["Borough"].str.replace("\n","")
df["Neighborhood"] = df["Neighborhood"].str.replace("\n","")

#Remove Boroughs that are 'Not assigned'
df = df[df['Borough'] != 'Not assigned']
df.head()

PostalCode           Borough                                  Neighborhood
2        M3A        North York                                     Parkwoods
3        M4A        North York                              Victoria Village
4        M5A  Downtown Toronto                    Regent Park / Harbourfront
5        M6A        North York             Lawrence Manor / Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government

Group neighborhoods with smilar postalCodes

In [70]:
df = df.groupby(["PostalCode", "Borough"])["Neighborhood"].apply(", ".join).reset_index()
df.head()

PostalCode      Borough                              Neighborhood
0        M1B  Scarborough                           Malvern / Rouge
1        M1C  Scarborough  Rouge Hill / Port Union / Highland Creek
2        M1E  Scarborough       Guildwood / Morningside / West Hill
3        M1G  Scarborough                                    Woburn
4        M1H  Scarborough                                 Cedarbrae

In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

In [71]:
df.shape

(103, 3)

### 3. Get geographical coordinates of the neighborhoods using the Geocoder package

In [72]:
df_geo_coor = pd.read_csv("http://cocl.us/Geospatial_data")
df_geo_coor.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Joining the dataframes df and df_geo_coor

In [73]:
df_toronto = pd.merge(df, df_geo_coor, how='left', left_on = 'PostalCode', right_on = 'Postal Code')
# remove the "Postal Code" column
df_toronto.drop("Postal Code", axis=1, inplace=True)
df_toronto.head()

PostalCode      Borough                              Neighborhood  \
0        M1B  Scarborough                           Malvern / Rouge   
1        M1C  Scarborough  Rouge Hill / Port Union / Highland Creek   
2        M1E  Scarborough       Guildwood / Morningside / West Hill   
3        M1G  Scarborough                                    Woburn   
4        M1H  Scarborough                                 Cedarbrae   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

### 4. Explore and cluster the neighborhoods in Toronto

#### 4.1. Get the latitude and longitude values of Toronto.

In [74]:
address = "Toronto, ON"

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto city are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto city are 43.6534817, -79.3839347.


#### 4.2. Create a map of the whole Toronto City with neighborhoods superimposed on top.

In [75]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)
map_toronto

In [76]:
# Add markers on the map

for lat, lng, borough, neighborhood in zip(
        df_toronto['Latitude'], 
        df_toronto['Longitude'], 
        df_toronto['Borough'], 
        df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  

map_toronto

##### 4.3. Map of a part of Toronto City
We are going to work with only the boroughs that contain the word "Toronto".

In [77]:
# "denc" = [D]owntown Toronto, [E]ast Toronto, [N]orth Toronto, [C]entral Toronto
df_toronto_denc = df_toronto[df_toronto['Borough'].str.contains("Toronto")].reset_index(drop=True)
df_toronto_denc.head()

PostalCode          Borough                     Neighborhood   Latitude  \
0        M4E     East Toronto                      The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West / Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar / The Beaches West  43.668999   
3        M4M     East Toronto                  Studio District  43.659526   
4        M4N  Central Toronto                    Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [78]:
# Plot again the map and the markers for this region.

map_toronto_denc = folium.Map(location=[latitude, longitude], zoom_start=12)
for lat, lng, borough, neighborhood in zip(
        df_toronto_denc['Latitude'], 
        df_toronto_denc['Longitude'], 
        df_toronto_denc['Borough'], 
        df_toronto_denc['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto_denc)  

map_toronto_denc

##### 4.4. Define Foursquare Credentials and Version
On the public repository on Github, I remove this field for the privacy!

In [80]:
CLIENT_ID = 'A2OFYYDNGKJ0SSM1Q4HU4QUNX3CIJ3ZHCEBA4110CBTEYQMS'
CLIENT_SECRET = 'YM5UGRHV5VXHMIK4QA3MDRUWK2K3WRHHC1WROSFG51ENSCR4'
VERSION = '20180605'

##### 4.5. Explore the first neighborhood in our data frame "df_toronto"

In [81]:
neighborhood_name = df_toronto_denc.loc[0, 'Neighborhood']
print(f"The first neighborhood's name is '{neighborhood_name}'.")

The first neighborhood's name is 'The Beaches'.


In [82]:
# Get the neighborhood's latitude and longitude values.
neighborhood_latitude = df_toronto_denc.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = df_toronto_denc.loc[0, 'Longitude'] # neighborhood longitude value

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.


Latitude and longitude values of The Beaches are 43.67635739999999, -79.2930312.

Now, let's get the top 100 venues that are in The Beaches within a radius of 500 meters.

In [83]:
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

# get the result to a json file
results = requests.get(url).json()

In [84]:
# Function that extracts the category of the venue

def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [86]:
# Now we are ready to clean the json and structure it into a pandas dataframe.

venues = results['response']['groups'][0]['items']
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues


C:\Users\suman\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  after removing the cwd from sys.path.


name         categories        lat        lng
0                   Glen Manor Ravine              Trail  43.676821 -79.293942
1  The Big Carrot Natural Food Market  Health Food Store  43.678879 -79.297734
2                 Grover Pub and Grub                Pub  43.679181 -79.297215
3                   Glen Stewart Park               Park  43.675278 -79.294647
4                       Upper Beaches       Neighborhood  43.680563 -79.292869

##### 4.6. Explore neighborhoods in a part of Toronto City
We are working on the data frame df_toronto_denc. Recall that, this region contain DENC of Toronto where,

"DENC" = [D]owntown Toronto, [E]ast Toronto, [N]orth Toronto, [C]entral Toronto

First, let's create a function to repeat the same process to all the neighborhoods in DENC of Toronto.

In [87]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    venues_list=[]
    
    for name, lat, lng in zip(names, latitudes, longitudes):
        # print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [88]:
# Now write the code to run the above function on each neighborhood and create a new dataframe called toronto_denc_venues

toronto_denc_venues = getNearbyVenues(names=df_toronto_denc['Neighborhood'],
                                   latitudes=df_toronto_denc['Latitude'],
                                   longitudes=df_toronto_denc['Longitude']
                                  )

In [89]:
toronto_denc_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                   Glen Stewart Park       43.675278       -79.294647   
4                       Upper Beaches       43.680563       -79.292869   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3               Park  
4       Neighborhood

In [90]:
# Let's check how many venues were returned for each neighborhood.

toronto_denc_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            55   
Brockton / Parkdale Village / Exhibition Place                         23   
Business reply mail Processing CentrE                                  16   
CN Tower / King and Spadina / Railway Lands / H...                     17   
Central Bay Street                                                     76   
Christie                                                               18   
Church and Wellesley                                                   81   
Commerce Court / Victoria Hotel                                       100   
Davisville                                                             35   
Davisville North                                                        9   
Dufferin / Dovercourt Village                                          18   
First Canadian Place / Underground city                               100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East / Union Station / Toronto Isl...                    100   
High Park / The Junction South                                         24   
India Bazaar / The Beaches West                                        19   
Kensington Market / Chinatown / Grange Park                            75   
Lawrence Park                                                           3   
Little Portugal / Trinity                                              50   
Moore Park / Summerhill East                                            3   
North Toronto West                                                     21   
Parkdale / Roncesvalles                                                13   
Queen's Park / Ontario Provincial Government                           34   
Regent Park / Harbourfront                                             45   
Richmond / Adelaide / King                                            100   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede / Swansea                                                    45   
St. James Town                                                        100   
St. James Town / Cabbagetown                                           43   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West / Rathnelly / South Hill / Fore...                     15   
The Annex / North Midtown / Yorkville                                  23   
The Beaches                                                             5   
The Danforth West / Riverdale                                          42   
Toronto Dominion Centre / Design Exchange                             100   
University of Toronto / Harbord                                        36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             55   
Brockton / Parkdale Village / Exhibition Place                          23   
Business reply mail Processing CentrE                                   16   
CN Tower / King and Spadina / Railway Lands / H...                      17   
Central Bay Street                                                      76   
Christie                                                                18   
Church and Wellesley                                                    81   
Commerce Court / Victoria Hotel                                        100   
Davisville                                                              35   
Davisville North                    

In [91]:
# Let's find out how many unique categories can be curated from all the returned venues
print('There are {} uniques categories.'.format(len(toronto_denc_venues['Venue Category'].unique())))

There are 236 uniques categories.


#### 4.7. Analyze Each Neighborhood

In [92]:
# one hot encoding
toronto_denc_onehot = pd.get_dummies(toronto_denc_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_denc_onehot['Neighborhood'] = toronto_denc_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_denc_onehot.columns[-1]] + list(toronto_denc_onehot.columns[:-1])
toronto_denc_onehot = toronto_denc_onehot[fixed_columns]

toronto_denc_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  ...  Theme Restaurant  Thrift / Vintage Store  \
0             0  ...                 0                       0   
1             0  ...                 0                       0   
2             0  ...                 0                       0   
3             0  ...                 0                       0   
4             0  ...                 0                       0   

   Toy / Game Store  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      1              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      0              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Women's Store  
0                 0                      0         0              0  
1                 0                      0         0              0  
2                 0                      0         0              0  
3                 0                      0         0              0  
4                 0                      0         0              0  

[5 rows x 236 columns]

In [93]:
# Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

toronto_denc_grouped = toronto_denc_onehot.groupby('Neighborhood').mean().reset_index()
toronto_denc_grouped.head()

Neighborhood  Yoga Studio  \
0                                        Berczy Park     0.000000   
1     Brockton / Parkdale Village / Exhibition Place     0.043478   
2              Business reply mail Processing CentrE     0.062500   
3  CN Tower / King and Spadina / Railway Lands / ...     0.000000   
4                                 Central Bay Street     0.013158   

   Afghan Restaurant   Airport  Airport Food Court  Airport Gate  \
0                0.0  0.000000            0.000000      0.000000   
1                0.0  0.000000            0.000000      0.000000   
2                0.0  0.000000            0.000000      0.000000   
3                0.0  0.058824            0.058824      0.058824   
4                0.0  0.000000            0.000000      0.000000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.000000         0.000000          0.000000             0.000000   
1        0.000000         0.000000          0.000000             0.000000   
2        0.000000         0.000000          0.000000             0.000000   
3        0.117647         0.176471          0.117647             0.000000   
4        0.000000         0.000000          0.000000             0.013158   

   ...  Theme Restaurant  Thrift / Vintage Store  Toy / Game Store  Trail  \
0  ...               0.0                     0.0               0.0    0.0   
1  ...               0.0                     0.0               0.0    0.0   
2  ...               0.0                     0.0               0.0    0.0   
3  ...               0.0                     0.0               0.0    0.0   
4  ...               0.0                     0.0               0.0    0.0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0                       0.018182               0.0   
1            0.0                       0.000000               0.0   
2            0.0                       0.000000               0.0   
3            0.0                       0.000000               0.0   
4            0.0                       0.013158               0.0   

   Vietnamese Restaurant  Wine Bar  Women's Store  
0                    0.0  0.000000            0.0  
1                    0.0  0.000000            0.0  
2                    0.0  0.000000            0.0  
3                    0.0  0.000000            0.0  
4                    0.0  0.013158            0.0  

[5 rows x 236 columns]

In [94]:
# Check the 10 most common venues in each neighborhood.

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_denc_grouped['Neighborhood']

for ind in np.arange(toronto_denc_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_denc_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1     Brockton / Parkdale Village / Exhibition Place                  Café   
2              Business reply mail Processing CentrE           Yoga Studio   
3  CN Tower / King and Spadina / Railway Lands / ...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                Bakery           Cheese Shop   
1        Breakfast Spot           Coffee Shop         Grocery Store   
2         Auto Workshop            Skate Park               Brewery   
3      Airport Terminal        Airport Lounge         Boat or Ferry   
4    Italian Restaurant          Burger Joint   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0                  Café            Restaurant     Seafood Restaurant   
1                   Gym             Pet Store  Performing Arts Venue   
2            Smoke Shop                   Spa         Farmers Market   
3      Sculpture Garden   Rental Car Location                  Plane   
4                  Café        Sandwich Place        Thai Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0              Beer Bar        Farmers Market       Greek Restaurant  
1             Nightclub    Italian Restaurant           Intersection  
2  Fast Food Restaurant            Restaurant          Burrito Place  
3              Boutique       Harbor / Marina           Airport Gate  
4  Gym / Fitness Center       Bubble Tea Shop         Ice Cream Shop

#### 4.8. Cluster neighborhoods
Run k-means to cluster the neighborhood into 5 clusters.

In [95]:
# set number of clusters
kclusters = 5

toronto_denc_grouped_clustering = toronto_denc_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_denc_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood

In [96]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_denc_merged = df_toronto_denc

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_denc_merged = toronto_denc_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_denc_merged.head() # check the last columns!

PostalCode          Borough                     Neighborhood   Latitude  \
0        M4E     East Toronto                      The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West / Riverdale  43.679557   
2        M4L     East Toronto  India Bazaar / The Beaches West  43.668999   
3        M4M     East Toronto                  Studio District  43.659526   
4        M4N  Central Toronto                    Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0                  Park                 Trail   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park      Sushi Restaurant   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               2                  Park           Swim School   

  3rd Most Common Venue   4th Most Common Venue 5th Most Common Venue  \
0                   Pub       Health Food Store   Dumpling Restaurant   
1    Italian Restaurant  Furniture / Home Store             Bookstore   
2           Coffee Shop           Movie Theater             Pet Store   
3             Gastropub                  Bakery               Brewery   
4              Bus Line             Event Space     Electronics Store   

         6th Most Common Venue        7th Most Common Venue  \
0                   Donut Shop  Eastern European Restaurant   
1               Ice Cream Shop                Grocery Store   
2                          Pub                Burrito Place   
3          American Restaurant                  Yoga Studio   
4  Eastern European Restaurant          Dumpling Restaurant   

     8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Doner Restaurant          Dance Studio                Dog Run  
1                      Pub           Pizza Place                 Lounge  
2               Restaurant          Liquor Store                Brewery  
3  Comfort Food Restaurant        Sandwich Place            Cheese Shop  
4               Donut Shop      Doner Restaurant                Dog Run

Finally, let's visualize the resulting clusters

In [97]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(
        toronto_denc_merged['Latitude'], 
        toronto_denc_merged['Longitude'], 
        toronto_denc_merged['Neighborhood'], 
        toronto_denc_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### 4.9. Examine Clusters
Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster.

Cluster 1

In [98]:
toronto_denc_merged.loc[toronto_denc_merged['Cluster Labels'] == 0, toronto_denc_merged.columns[[1] + list(range(5, toronto_denc_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0       East Toronto               0                  Park   
1       East Toronto               0      Greek Restaurant   
2       East Toronto               0                  Park   
3       East Toronto               0                  Café   
5    Central Toronto               0        Sandwich Place   
6    Central Toronto               0        Clothing Store   
7    Central Toronto               0           Pizza Place   
9    Central Toronto               0           Coffee Shop   
11  Downtown Toronto               0           Coffee Shop   
12  Downtown Toronto               0   Japanese Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   
17  Downtown Toronto               0           Coffee Shop   
18  Downtown Toronto               0                  Café   
19  Downtown Toronto               0           Coffee Shop   
20  Downtown Toronto               0           Coffee Shop   
21  Downtown Toronto               0           Coffee Shop   
24   Central Toronto               0                  Café   
25  Downtown Toronto               0                  Café   
26  Downtown Toronto               0                  Café   
27  Downtown Toronto               0       Airport Service   
28  Downtown Toronto               0           Coffee Shop   
29  Downtown Toronto               0           Coffee Shop   
30  Downtown Toronto               0         Grocery Store   
31      West Toronto               0              Pharmacy   
32      West Toronto               0                   Bar   
33      West Toronto               0                  Café   
34      West Toronto               0    Mexican Restaurant   
35      West Toronto               0             Gift Shop   
36      West Toronto               0                  Café   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0           Yoga Studio   

   2nd Most Common Venue  3rd Most Common Venue   4th Most Common Venue  \
0                  Trail                    Pub       Health Food Store   
1            Coffee Shop     Italian Restaurant  Furniture / Home Store   
2       Sushi Restaurant            Coffee Shop           Movie Theater   
3            Coffee Shop              Gastropub                  Bakery   
5      Convenience Store      Food & Drink Shop        Asian Restaurant   
6            Coffee Shop            Yoga Studio     Sporting Goods Shop   
7         Sandwich Place           Dessert Shop                    Café   
9                    Pub    American Restaurant              Restaurant   
11           Pizza Place     Italian Restaurant                     Pub   
12           Coffee Shop                Gay Bar        Sushi Restaurant   
13                  Park                 Bakery                     Pub   
14        Clothing Store        Bubble Tea Shop     Japanese Restaurant   
15                  Café             Restaurant                   Diner   
16          Cocktail Bar                 Bakery             Cheese Shop   
17    Italian Restaurant           Burger Joint     Japanese Restaurant   
18           Coffee Shop             Restaurant                  Bakery   
19              Aquarium     Italian Restaurant                   Hotel   
20                  Café                  Hotel              Restaurant   
21                  Café             Restaurant                   Hotel   
24           Coffee Shop         Sandwich Place           Grocery Store   
25   Japanese Restaurant                    Bar               Bookstore   
26           Coffee Shop  Vietnamese Restaurant                     Bar   
27      Airport Terminal         Airport Lounge           Boat or Ferry   
28    Seafood Restaurant             Resta

**Cluster 2**

In [99]:
toronto_denc_merged.loc[toronto_denc_merged['Cluster Labels'] == 1, toronto_denc_merged.columns[[1] + list(range(5, toronto_denc_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22         Women's Store         Deli / Bodega  Ethiopian Restaurant   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
22     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22            Donut Shop      Doner Restaurant                Dog Run

**Cluster 3**

In [100]:
toronto_denc_merged.loc[toronto_denc_merged['Cluster Labels'] == 2, toronto_denc_merged.columns[[1] + list(range(5, toronto_denc_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
4    Central Toronto               2                  Park   
10  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4            Swim School              Bus Line           Event Space   
10            Playground                 Trail         Deli / Bodega   

   5th Most Common Venue        6th Most Common Venue  \
4      Electronics Store  Eastern European Restaurant   
10  Ethiopian Restaurant            Electronics Store   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
4           Dumpling Restaurant            Donut Shop      Doner Restaurant   
10  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   

   10th Most Common Venue  
4                 Dog Run  
10       Doner Restaurant

**Cluster 4**

In [101]:
toronto_denc_merged.loc[toronto_denc_merged['Cluster Labels'] == 3, toronto_denc_merged.columns[[1] + list(range(5, toronto_denc_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
23  Central Toronto               3         Jewelry Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23                 Trail    Mexican Restaurant      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23         Women's Store      Department Store     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

**Cluster 5**

In [102]:
toronto_denc_merged.loc[toronto_denc_merged['Cluster Labels'] == 4, toronto_denc_merged.columns[[1] + list(range(5, toronto_denc_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
8  Central Toronto               4            Playground   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8          Tennis Court            Restaurant         Women's Store   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
8         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8   Dumpling Restaurant            Donut Shop       Doner Restaurant